File: src/bayes_poker/player_metrics/enums.py
"""玩家指标模块的枚举类型定义。

移植自 G5.Logic/Enums.cs，仅保留玩家指标构建所需的枚举。
"""

from __future__ import annotations

from enum import IntEnum, auto


class TableType(IntEnum):
    """桌型枚举。

    值等于该桌型的最大玩家数。
    """

    HEADS_UP = 2
    SIX_MAX = 6


class Street(IntEnum):
    """游戏阶段（街）。"""

    UNKNOWN = 0
    PREFLOP = auto()
    FLOP = auto()
    TURN = auto()
    RIVER = auto()


class Position(IntEnum):
    """玩家位置。

    索引顺序与 G5.Logic 保持一致：SB=0, BB=1, ...
    """

    SMALL_BLIND = 0
    BIG_BLIND = 1
    UTG = 2
    HJ = 3  # Hijack (MP in 6-max)
    CO = 4  # Cutoff
    BUTTON = 5
    EMPTY = 6


class ActionType(IntEnum):
    """玩家动作类型。"""

    FOLD = 0
    CHECK = auto()
    CALL = auto()
    BET = auto()
    RAISE = auto()
    ALL_IN = auto()
    WINS = auto()
    MONEY_RETURNED = auto()
    NO_ACTION = auto()

    @property
    def is_raise_action(self) -> bool:
        """是否为加注类动作（Bet/Raise/AllIn）。"""
        return self in (ActionType.BET, ActionType.RAISE, ActionType.ALL_IN)

    @property
    def is_valid_action(self) -> bool:
        """是否为有效的博弈动作（排除 Wins/MoneyReturned/NoAction）。"""
        return self in (
            ActionType.FOLD,
            ActionType.CHECK,
            ActionType.CALL,
            ActionType.BET,
            ActionType.RAISE,
            ActionType.ALL_IN,
        )

File: crates/poker_stats_rs/src/enums.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum TableType {
    HeadsUp = 2,
    SixMax = 6,
}

impl TableType {
    pub fn from_u8(v: u8) -> Self {
        match v {
            2 => TableType::HeadsUp,
            _ => TableType::SixMax,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum Street {
    PreFlop = 1,
    Flop = 2,
    Turn = 3,
    River = 4,
}

impl Street {
    pub fn from_str(s: &str) -> Self {
        match s.to_lowercase().as_str() {
            "preflop" => Street::PreFlop,
            "flop" => Street::Flop,
            "turn" => Street::Turn,
            "river" => Street::River,
            _ => Street::PreFlop,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum Position {
    SmallBlind = 0,
    BigBlind = 1,
    UTG = 2,
    HJ = 3,
    CutOff = 4,
    Button = 5,
}

impl Position {
    pub fn from_index(idx: usize, total: usize) -> Self {
        if total == 2 {
            if idx == 0 { Position::SmallBlind } else { Position::BigBlind }
        } else {
            match idx {
                0 => Position::SmallBlind,
                1 => Position::BigBlind,
                i if i == total - 1 => Position::Button,
                i if i == total - 2 => Position::CutOff,
                i if i == total - 3 => Position::HJ,
                _ => Position::UTG,
            }
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum ActionType {
    Fold = 0,
    Check = 1,
    Call = 2,
    Bet = 3,
    Raise = 4,
    AllIn = 5,
}

impl ActionType {
    pub fn from_str(s: &str) -> Self {
        match s.to_lowercase().as_str() {
            "fold" => ActionType::Fold,
            "check" => ActionType::Check,
            "call" => ActionType::Call,
            "bet" => ActionType::Bet,
            "raise" => ActionType::Raise,
            "allin" | "all_in" | "all-in" => ActionType::AllIn,
            _ => ActionType::Fold,
        }
    }
    
    pub fn is_raise_action(&self) -> bool {
        matches!(self, ActionType::Bet | ActionType::Raise | ActionType::AllIn)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
#[repr(u8)]
pub enum PreflopPotType {
    Limped = 0,
    #[default]
    SingleRaised = 1,
    ThreeBetPlus = 2,
}

impl PreflopPotType {
    pub fn from_raise_count(count: i32) -> Self {
        match count {
            0 => PreflopPotType::Limped,
            1 => PreflopPotType::SingleRaised,
            _ => PreflopPotType::ThreeBetPlus,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum BetSizingCategory {
    Bet0To40 = 0,
    Bet40To80 = 1,
    Bet80To120 = 2,
    BetOver120 = 3,
}

impl BetSizingCategory {
    pub fn from_pot_percentage(pct: f64) -> Self {
        if pct < 0.40 {
            BetSizingCategory::Bet0To40
        } else if pct < 0.80 {
            BetSizingCategory::Bet40To80
        } else if pct < 1.20 {
            BetSizingCategory::Bet80To120
        } else {
            BetSizingCategory::BetOver120
        }
    }
}
